File: sdks/php/src/Api/Emails.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\BatchSendRequest;
use Apollo\Deploy\Signal\Sdk\BatchSendResponse;
use Apollo\Deploy\Signal\Sdk\BulkCancelResponse;
use Apollo\Deploy\Signal\Sdk\CancelResponse;
use Apollo\Deploy\Signal\Sdk\EmailDetailResponse;
use Apollo\Deploy\Signal\Sdk\SendEmailRequest;
use Apollo\Deploy\Signal\Sdk\SendEmailResponse;
use Apollo\Deploy\Signal\Sdk\StreamEventResponse;
use Apollo\Deploy\Signal\Sdk\StreamTokenResponse;
use Apollo\Deploy\Signal\Sdk\ValidateLinksRequest;
use Apollo\Deploy\Signal\Sdk\ValidateLinksResponse;
use Psr\Http\Message\ResponseInterface;

/** EmailsAPI domain operations. */
class Emails
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param SendEmailRequest $input
     * @return SendEmailResponse
     * @throws SDKError
     */
    public function sendEmail(
        SendEmailRequest $input,
    ): SendEmailResponse
    {
        $path = '/v1/emails';

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(SendEmailResponse::class, 'fromArray')) {
            return SendEmailResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $emailId
     * @return EmailDetailResponse
     * @throws SDKError
     */
    public function getEmail(
        string $emailId,
    ): EmailDetailResponse
    {
        $path = '/v1/emails/${emailId}';
        $path = str_replace('${emailId}', rawurlencode(Transport::wireValue($emailId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(EmailDetailResponse::class, 'fromArray')) {
            return EmailDetailResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param BatchSendRequest $input
     * @return BatchSendResponse
     * @throws SDKError
     */
    public function batchSendEmails(
        BatchSendRequest $input,
    ): BatchSendResponse
    {
        $path = '/v1/emails/batch';

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(BatchSendResponse::class, 'fromArray')) {
            return BatchSendResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $emailId
     * @return CancelResponse
     * @throws SDKError
     */
    public function cancelEmail(
        string $emailId,
    ): CancelResponse
    {
        $path = '/v1/emails/${emailId}';
        $path = str_replace('${emailId}', rawurlencode(Transport::wireValue($emailId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(CancelResponse::class, 'fromArray')) {
            return CancelResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @return BulkCancelResponse
     * @throws SDKError
     */
    public function bulkCancelEmails(
    ): BulkCancelResponse
    {
        $path = '/v1/emails';

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(BulkCancelResponse::class, 'fromArray')) {
            return BulkCancelResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @return StreamTokenResponse
     * @throws SDKError
     */
    public function issueStreamToken(
        string $projectId,
    ): StreamTokenResponse
    {
        $path = '/v1/emails/${projectId}/stream/token';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(StreamTokenResponse::class, 'fromArray')) {
            return StreamTokenResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @return \Generator<int, SSEEvent>
     * @throws SDKError
     */
    public function streamEmailEvents(
    ): \Generator
    {
        $path = '/v1/emails/stream';

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
            'stream' => true,
        ];

        $response = $this->transport->executeRequest($request);
        // Transport::streamEvents enforces Transport::MAX_SSE_LINE_BYTES and
        // Transport::MAX_RESPONSE_BODY_BYTES while parsing the response.
        foreach ($this->transport->streamEvents($response) as $event) {
            yield $event;
        }
    }

    /**
     * @param ValidateLinksRequest $input
     * @return ValidateLinksResponse
     * @throws SDKError
     */
    public function validateLinks(
        ValidateLinksRequest $input,
    ): ValidateLinksResponse
    {
        $path = '/v1/emails/validate';

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ValidateLinksResponse::class, 'fromArray')) {
            return ValidateLinksResponse::fromArray($data);
        }
        return $data;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/emails.rs
use crate::error::SdkError;
use crate::transport::Transport;
use crate::transport::SseEventStream;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ BatchSendRequest, BatchSendResponse, BulkCancelResponse, CancelResponse, EmailDetailResponse, SendEmailRequest, SendEmailResponse, StreamEventResponse, StreamTokenResponse, ValidateLinksRequest, ValidateLinksResponse };

/// EmailsAPI API operations.
pub struct EmailsAPI {
    transport: Transport,
}

impl EmailsAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn send_email(
        &self,
        body: &SendEmailRequest,
    ) -> Result<SendEmailResponse, SdkError> {
        let mut path = "/v1/emails".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<SendEmailResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn get_email(
        &self,
        email_id: String,
    ) -> Result<EmailDetailResponse, SdkError> {
        let mut path = "/v1/emails/${emailId}".to_string();
        path = path.replace("${emailId}", &encode_path(&email_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<EmailDetailResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn batch_send_emails(
        &self,
        body: &BatchSendRequest,
    ) -> Result<BatchSendResponse, SdkError> {
        let mut path = "/v1/emails/batch".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<BatchSendResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn cancel_email(
        &self,
        email_id: String,
    ) -> Result<CancelResponse, SdkError> {
        let mut path = "/v1/emails/${emailId}".to_string();
        path = path.replace("${emailId}", &encode_path(&email_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<CancelResponse>(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn bulk_cancel_emails(
        &self,
    ) -> Result<BulkCancelResponse, SdkError> {
        let mut path = "/v1/emails".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<BulkCancelResponse>(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn issue_stream_token(
        &self,
        project_id: String,
    ) -> Result<StreamTokenResponse, SdkError> {
        let mut path = "/v1/emails/${projectId}/stream/token".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<StreamTokenResponse>(
            Method::POST,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn stream_email_events(
        &self,
    ) -> Result<SseEventStream<StreamEventResponse>, SdkError> {
        let mut path = "/v1/emails/stream".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_sse::<StreamEventResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn validate_links(
        &self,
        body: &ValidateLinksRequest,
    ) -> Result<ValidateLinksResponse, SdkError> {
        let mut path = "/v1/emails/validate".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ValidateLinksResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
